presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusSplitManager.java

```diff
@@ -0,0 +1,162 @@
+        implements ConnectorSplitManager
+{
+    static final long DAYS_IN_A_YEAR = 365;
+    static final long DAYS_IN_A_WEEK = 7;
+    static final long MILLIS_IN_A_SECOND = 1000;
+    private final PrometheusClient prometheusClient;
+
+    @Inject
+    public PrometheusSplitManager(PrometheusClient prometheusClient)
+    {
+        this.prometheusClient = requireNonNull(prometheusClient, "client is null");
+        requireNonNull(prometheusClient.config.getQueryChunkSizeDuration(), "query-chunk-size-duration is null");
+        requireNonNull(prometheusClient.config.getMaxQueryRangeDuration(), "max-query-range-duration is null");
+    }
+
+    @Override
+    public ConnectorSplitSource getSplits(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorTableHandle connectorTableHandle,
+            SplitSchedulingStrategy splitSchedulingStrategy)
+    {
+        PrometheusTableHandle tableHandle = (PrometheusTableHandle) connectorTableHandle;
+        PrometheusTable table = prometheusClient.getTable(tableHandle.getSchemaName(), tableHandle.getTableName());
+
+        // this can happen if table is removed during a query
+        if (table == null) {
+            throw new TableNotFoundException(tableHandle.toSchemaTableName());
+        }
+        List<ConnectorSplit> splits = new ArrayList<>();
+        generateTimesForSplits().forEach(time -> {
+            try {
+                splits.add(new PrometheusSplit(buildQuery(prometheusClient.config.getPrometheusURI(), time, table.getName())));
+            }
+            catch (URISyntaxException e) {
+                throw new UnsupportedOperationException(e.getMessage());
+            }
+        });
+        return new FixedSplitSource(splits);
+    }
+
+    // URIBuilder handles URI encode
+    private URI buildQuery(URI baseURI, String time, String tableName)
+            throws URISyntaxException
+    {
+        List<NameValuePair> nameValuePairs = new ArrayList<>(2);
+        nameValuePairs.add(new BasicNameValuePair("query",
+                tableName + "[" + prometheusClient.config.getQueryChunkSizeDuration() + "]"));
+        nameValuePairs.add(new BasicNameValuePair("time",
+                time));
+        return new URIBuilder(baseURI.toString())
+                .setPath("api/v1/query")
+                .setParameters(nameValuePairs).build();
+    }
+
+    /**
+     * Get the start times in seconds that divide up the query using the config value:
+     * maxQueryRangeDuration
+     * queryChunkSizeDuration
+     * The times will be used in queries to Prometheus like: `http://localhost:9090/api/v1/query?query=up[3w]&time=1568229904"`
+     * ** NOTE: Prometheus instant query wants the end time specified, you get duration before that time returned
+     *
+     * @return list of start times as epoch seconds, like ["1568053244", "1568926595"]
+     */
+    private List<String> generateTimesForSplits()
```

Comment:
Definitely need tests for this. We need to make sure we're not querying overlapping periods of time.

Reply:
We have 4 tests now that indirectly test this function, they look at the query generated including evaluating whether the 1st chunk has the right end time; the last chunk has the right end time; the spacing of chunk end times is correct and whether the last end time is still correct if the duration and chunk times don't divide evenly:

TestPrometheusSplit#testQueryDividedIntoSplitsFirstSplitHasRightTime
TestPrometheusSplit#testQueryDividedIntoSplitsLastSplitHasRightTime
TestPrometheusSplit#testQueryDividedIntoSplitsShouldHaveCorrectSpacingBetweenTimes
TestPrometheusSplit#testQuerySplitsCorrectWithNonModuloZeroDurationDividedByChunk

I'm up for writing some more if we can identify some other likely mistakes...?